In [1]:
import matplotlib.pyplot as plt
import numpy as np
import dask_geopandas as dg

# ATTENTION:
User must follow below instructions and replace the AWS access key and secret access key. This is crucial for the analysis to work.

##### Utilizing data from: https://source.coop/repositories/wherobots/usa-structures/description

All data on Source Cooperative, are hosted on AWS S3 bucket. In order to access them, you need credentials that you can generate on Source Cooperative website. Atfer logging in, click on your name at the top right corner, and then click on your username. Then navigate to "Manage" page on the left side. At the bottom of this page you will find a section called "API Keys". If no key has been generated before, generate a new one and then copy the values for each of the following keys, and paste them in the following cell.

source.coop website: https://source.coop/

###### Source: https://github.com/github.com/HamedAlemo/vector-data-tutorial/scalable_vector_analysis.ipynb

In [2]:
##################################
#   Read Above 'ATTENTION' Note  #
##################################

AWS_ACCESS_KEY_ID = "<YOUR ACCESS KEY>"
AWS_SECRET_ACCESS_KEY = "<YOUR SECRET ACCESS KEY>"

In [3]:
import boto3
s3_client = boto3.client('s3',
                         aws_access_key_id = AWS_ACCESS_KEY_ID, 
                         aws_secret_access_key = AWS_SECRET_ACCESS_KEY,
                         endpoint_url='https://data.source.coop'
                        )

In [4]:
from dask.distributed import Client, LocalCluster
cluster = LocalCluster()
client = Client(cluster)
print(client.dashboard_link)

http://127.0.0.1:8787/status


##### Local path for downloading the data

If running from analysis file within the 'saved'//mounted folder and do NOT wish to save the raw data - delete the
first local_path and uncomment (remove #) on the second one

In [ ]:
local_path = "./data/" # saves data to machine
# local_path ="/data/"

In [7]:
import utils as ut

Install time will vary. As of 12/9/2024, the dataset is 10 geoparquets each equating to ~2 Gigabytes. This took ~40 minutes to install on my personal machine:

~ 3 Yr Old Laptop

Windows 11

16 Gb RAM (Docker is limited to 12 Gb)

Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz, 2592 Mhz, 6 Core(s), 12 Logical Processor(s)

In [7]:
structure_ddf = ut.get_US_structures_all(s3_client, local_path, blocksize = "16M") #256M is regular block size

File already exists locally. No download needed.
File already exists locally. No download needed.
File already exists locally. No download needed.
File already exists locally. No download needed.
File already exists locally. No download needed.
File already exists locally. No download needed.
File already exists locally. No download needed.
File already exists locally. No download needed.
File already exists locally. No download needed.
File already exists locally. No download needed.


In [8]:
structure_ddf.head(5)

geometry  BUILD_ID       OCC_CLS  \
0  MULTIPOLYGON (((-170.8292 -14.32645, -170.8292...     12059  Unclassified   
1  MULTIPOLYGON (((-170.82798 -14.32867, -170.827...     29894  Unclassified   
2  MULTIPOLYGON (((-170.82786 -14.3283, -170.8278...     29886  Unclassified   
3  MULTIPOLYGON (((-170.82747 -14.32844, -170.827...     29879  Unclassified   
4  MULTIPOLYGON (((-170.82714 -14.32846, -170.827...     29885  Unclassified   

       PRIM_OCC SEC_OCC PROP_ADDR PROP_CITY         PROP_ST PROP_ZIP OUTBLDG  \
0  Unclassified    None      None      None  American Samoa     None    None   
1  Unclassified    None      None      None  American Samoa     None    None   
2  Unclassified    None      None      None  American Samoa     None    None   
3  Unclassified    None      None      None  American Samoa     None    None   
4  Unclassified    None      None      None  American Samoa     None    None   

   ...                USNG   LONGITUDE   LATITUDE              IMAGE_NAME  \
0  ...  02L NK 18412 16164 -170.829258 -14.326434        104001005B4C3F00   
1  ...  02L NK 18540 15909 -170.828069 -14.328742  NOAA Topographic LiDAR   
2  ...  02L NK 18569 15948 -170.827797 -14.328382  NOAA Topographic LiDAR   
3  ...  02L NK 18594 15936 -170.827566 -14.328497  NOAA Topographic LiDAR   
4  ...  02L NK 18636 15935 -170.827176 -14.328503  NOAA Topographic LiDAR   

   IMAGE_DATE  VAL_METHOD REMARKS                                    UUID  \
0  2020-05-14   Automated    None  {245aca01-8d02-4340-810e-3872ee35e7ef}   
1  2013-01-01  Unverified    None  {a46fac03-1829-47a2-b63f-86d4e52700a8}   
2  2013-01-01  Unverified    None  {4c8bb4a4-86e1-4cb2-908c-a367e7dfa525}   
3  2013-01-01  Unverified    None  {22106c1f-562e-4468-a496-32d407fd8b4f}   
4  2013-01-01  Unverified    None  {bc01e591-8516-49bf-b66e-dd21667de2f7}   

                                                bbox          geohash  
0  {'xmin': -170.82931163799998, 'ymin': -14.3264...  2jqw2xvymz0vch1  
1  {'xmin': -170.82815531799997, 'ymin': -14.3288...  2jqw2xwc12zkvd0  
2  {'xmin': -170.82790860099996, 'ymin': -14.3284...  2jqw2xwgpg8y3ej  
3  {'xmin': -170.82767668499997, 'ymin': -14.3286...  2jqw2xx4mn23x0c  
4  {'xmin': -170.82721784799998, 'ymin': -14.3285...  2jqw2xx6qm2znys  

[5 rows x 29 columns]

In [8]:
structure_ddf.tail(5)

geometry  BUILD_ID  \
539723  MULTIPOLYGON (((173.18076 52.83347, 173.1808 5...   7000006   
539724  MULTIPOLYGON (((173.18831 52.83659, 173.18866 ...   7000009   
539725  MULTIPOLYGON (((174.09103 52.7369, 174.09113 5...   8000048   
539726  MULTIPOLYGON (((174.09056 52.73711, 174.09132 ...   8000047   
539727  MULTIPOLYGON (((174.10244 52.73433, 174.10273 ...   8000036   

             OCC_CLS                 PRIM_OCC SEC_OCC PROP_ADDR PROP_CITY  \
539723  Unclassified             Unclassified    None      None      None   
539724  Unclassified             Unclassified    None      None      None   
539725    Government  Non-Civilian Structures    None      None      None   
539726    Government  Non-Civilian Structures    None      None      None   
539727    Government  Non-Civilian Structures    None      None      None   

       PROP_ST PROP_ZIP OUTBLDG  ...                USNG   LONGITUDE  \
539723  Alaska     None    None  ...  59U PU 46903 55967  173.180739   
539724  Alaska     None    None  ...  59U PU 47405 56303  173.188346   
539725  Alaska     None    None  ...  60U UD 03622 46972  174.091134   
539726  Alaska     None    None  ...  60U UD 03627 47013  174.091178   
539727  Alaska     None    None  ...  60U UD 04389 46663  174.102666   

         LATITUDE     IMAGE_NAME  IMAGE_DATE VAL_METHOD REMARKS  \
539723  52.833410  NOT AVAILABLE  2018-07-03     Manual    None   
539724  52.836292  NOT AVAILABLE  2018-07-03     Manual    None   
539725  52.736898  NOT AVAILABLE  2018-07-03     Manual    None   
539726  52.737275  NOT AVAILABLE  2018-07-03     Manual    None   
539727  52.734406  NOT AVAILABLE  2018-07-03     Manual    None   

                                          UUID  \
539723  {a9045bdc-dab4-4de6-b49b-58eab8581e25}   
539724  {29993d4d-b59a-4d5c-9567-4c51b12059dc}   
539725  {17e345d8-b90d-439d-ad89-7984758ca934}   
539726  {a74622e4-0bc2-4809-9479-e77c6d24abc4}   
539727  {17e6eb27-63de-41f4-823b-fa4ff6f4851d}   

                                                     bbox          geohash  
539723  {'xmin': 173.1806779100001, 'ymin': 52.8333547...  zc7hsf42gtw3n05  
539724  {'xmin': 173.1880302520001, 'ymin': 52.8359877...  zc7hsfxcsvvefwt  
539725  {'xmin': 174.09102922500006, 'ymin': 52.736834...  zc7u183q9w2fn6e  
539726  {'xmin': 174.09055533200012, 'ymin': 52.736969...  zc7u1892ghk2u76  
539727  {'xmin': 174.10243806300002, 'ymin': 52.734255...  zc7u1b18nn6e3h5  

[5 rows x 29 columns]

### Verify Data Loaded Correctly
Specifically looking for data that will show a different 'PROP_ST' column besides American Samoa and Alaska (the .head and .tail for the structure_ddf table)

In [17]:
partition40 = structure_ddf.partitions[40].compute()

In [21]:
partition40.head(5)

geometry  BUILD_ID      OCC_CLS  \
0  MULTIPOLYGON (((-111.71603 40.37411, -111.7161...   6356322  Residential   
1  MULTIPOLYGON (((-111.7162 40.37468, -111.71619...   6356424  Residential   
2  MULTIPOLYGON (((-111.71587 40.37469, -111.7158...   6356420  Residential   
3  MULTIPOLYGON (((-111.71567 40.37345, -111.7156...   6356164  Residential   
4  MULTIPOLYGON (((-111.71552 40.37399, -111.7155...   6356225  Residential   

                 PRIM_OCC SEC_OCC             PROP_ADDR       PROP_CITY  \
0  Single Family Dwelling    None  1000 NORTH 1050 EAST  PLEASANT GROVE   
1  Single Family Dwelling    None  1053 EAST 1010 NORTH  PLEASANT GROVE   
2  Single Family Dwelling    None  1061 EAST 1010 NORTH  PLEASANT GROVE   
3  Single Family Dwelling    None   933 NORTH 1100 EAST  PLEASANT GROVE   
4  Single Family Dwelling    None   967 NORTH 1100 EAST  PLEASANT GROVE   

  PROP_ST PROP_ZIP OUTBLDG  ...                USNG   LONGITUDE   LATITUDE  \
0    Utah    84062    None  ...  12T VK 39208 69537 -111.716103  40.374204   
1    Utah    84062    None  ...  12T VK 39188 69584 -111.716341  40.374622   
2    Utah    84062    None  ...  12T VK 39218 69584 -111.715985  40.374631   
3    Utah    84062    None  ...  12T VK 39249 69468 -111.715610  40.373586   
4    Utah    84062    None  ...  12T VK 39251 69504 -111.715594  40.373907   

         IMAGE_NAME  IMAGE_DATE VAL_METHOD REMARKS  \
0  105001001D0AA300  2020-05-04  Automated    None   
1  105001001D0AA300  2020-05-04  Automated    None   
2  105001001D0AA300  2020-05-04  Automated    None   
3  105001001D0AA300  2020-05-04  Automated    None   
4  105001001D0AA300  2020-05-04  Automated    None   

                                     UUID  \
0  {94234441-bcf3-4b26-b88b-54b488b7e794}   
1  {ba007b62-1bbe-4538-9d94-ee5a347a3041}   
2  {4ccca22c-7705-4892-8da4-e00cb91d7f80}   
3  {76039646-81e2-4a9d-a77a-70512feb8bac}   
4  {e06859b7-dd6e-42d3-8f92-dcae99305093}   

                                                bbox          geohash  
0  {'xmin': -111.71617885399996, 'ymin': 40.37411...  9x0t9vfuf15vqc3  
1  {'xmin': -111.71650439999996, 'ymin': 40.37456...  9x0t9vfxhrj4g6r  
2  {'xmin': -111.71610170499997, 'ymin': 40.37457...  9x0t9vfzkbsn56z  
3  {'xmin': -111.71566964299996, 'ymin': 40.37345...  9x0t9vg1jwgyz1x  
4  {'xmin': -111.71568185699999, 'ymin': 40.37381...  9x0t9vg5jbccvg3  

[5 rows x 29 columns]

In [19]:
partition40.shape

(873670, 29)

Each partition has 16mb of data, roughly 873,670 rows. This will vary based on the blocksize chosen when calling the ut.get_US_structures_all() function.

### Coordinate Reference System (CRS)
Since this dataset captures "every structure larger than 450 square feet" we have a base minimum, however, to get specific building sizes, or plot, we will need to make sure the CRS is for the specific region/area you are analyzing. We will re-project based on the 'geometries' column.

First, we will check to see if we need to project to a CRS for the first time, or re-project the current one. To check if the data has a CRS, the following code will suffice.

In [31]:
structure_ddf.crs

<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: World.
- bounds: (-180.0, -90.0, 180.0, 90.0)
Datum: World Geodetic System 1984 ensemble
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

For all users, you will need to re-project. Research online to ensure you use the correct EPSG value for your area.

### Analyze Statewide Massachusetts data
Massachusetts can use EPSG:6492, this is in US Survey feet which means we don't need to worry about converting from meters to feet. 

NOTE: I do not adjust the US survey feet to International feet.

In [8]:
massachusetts_ddf = structure_ddf[structure_ddf['PROP_ST'] == 'Massachusetts']

In [9]:
massachusetts_ddf = massachusetts_ddf.to_crs(epsg=2249) # can also try 6492 but City of Worcester shapefile uses 2249

Calculate the square footage for each building and make this stored in a new column 'structure_size'.

In [10]:
massachusetts_ddf['structure_size'] = massachusetts_ddf.geometry.area

### Checking the validity of the 'PROP_CITY' column

##### Compare the number of rows when using 'PROP_CITY' == 'Worcester' VS a shapefile of Worcester as a Spatial Filter and finding intersections
(Using massachusetts_ddf ensures not grabbing Worcester or Boston city data for other states)

We will use .shape to compare the number of rows

To ensure we use the correct boundaries for Worcester, we will download a shapefile of the layout. This will be downloaded to the 'data' folder and not done in an automatic coding format. However, if you find a different shapefile/city, the same approach can be used. Just mount a folder when you run the docker container from a directory that has the associated city boundary shape files, then you can change the 'dg_read_file' filepath. (Future Implementation: Use selenium to download this file within docker so that users don't need to have data downloaded with the Docker Image)

https://opendata.worcesterma.gov/datasets/worcesterma::city-boundary-1/about

In [11]:
worcester_PROP_CITY_ddf = massachusetts_ddf[massachusetts_ddf['PROP_CITY'] == 'Worcester']

Confirm we are using the crs defined for massachusetts_ddf:

In [13]:
worcester_PROP_CITY_ddf.crs

<Projected CRS: EPSG:2249>
Name: NAD83 / Massachusetts Mainland (ftUS)
Axis Info [cartesian]:
- X[east]: Easting (US survey foot)
- Y[north]: Northing (US survey foot)
Area of Use:
- name: United States (USA) - Massachusetts onshore - counties of Barnstable; Berkshire; Bristol; Essex; Franklin; Hampden; Hampshire; Middlesex; Norfolk; Plymouth; Suffolk; Worcester.
- bounds: (-73.5, 41.46, -69.86, 42.89)
Coordinate Operation:
- name: SPCS83 Massachusetts Mainland zone (US survey foot)
- method: Lambert Conic Conformal (2SP)
Datum: North American Datum 1983
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich

In [13]:
worcester_boundary = dg.read_file(f"./data/Worcester/City_Boundary.shp", chunksize = 75000) # chunksize specifies number of rows per chunk 

# Repeat from above: 
# If you find a different shapefile/city, the same approach can be used. 
# Just mount a folder when you run the docker container from a directory that has the associated city boundary shape files, 
# then you can change the 'dg.read_file' filepath.

# It is suggested that you put the data in a data folder within the same directory!

##### If the above line of code gave you an error similar to:

ERROR 1: PROJ: proj_create_from_database: Open of /opt/conda/envs/us_structures_analysis/share/proj failed

Try re-running it

or

Copy the folder that contains all of the shapefile files outside the mounted folder, so it is in the main directory.

For instance. I use a mounted 'saved' folder, with a 'data' folder inside that contains all the geoparquets plus a 'Worcester' folder which contains all the files for the Worcester shapefile. I copied the 'Worcester' folder so that it is outside of the 'saved' folder.

In [14]:
worcester_boundary.crs
# You want to follow the crs of the boundary/shapefile you are using, make sure the dask dataframe you are using matches this

<Projected CRS: EPSG:2249>
Name: NAD83 / Massachusetts Mainland (ftUS)
Axis Info [cartesian]:
- X[east]: Easting (US survey foot)
- Y[north]: Northing (US survey foot)
Area of Use:
- name: United States (USA) - Massachusetts onshore - counties of Barnstable; Berkshire; Bristol; Essex; Franklin; Hampden; Hampshire; Middlesex; Norfolk; Plymouth; Suffolk; Worcester.
- bounds: (-73.5, 41.46, -69.86, 42.89)
Coordinate Operation:
- name: SPCS83 Massachusetts Mainland zone (US survey foot)
- method: Lambert Conic Conformal (2SP)
Datum: North American Datum 1983
- Ellipsoid: GRS 1980
- Prime Meridian: Greenwich

#### Since we are comparing the entire data set to the state/city column effectiveness (PROP_ST and PROP_CITY), we must make the entire dataset be the same crs as our boundary/shapefile

##### ** BEFORE WE ONLY CHANGED THE MASSACHUSETTS_DDF CRS **

OVERVIEW:
- This is testing the usefulness of the datasets columns for specifying if a structure is in a specific state/city.



### If you go on to review another state. You may need to re-project the CRS for the entire dataset. The following is a good site for visually seeing what the EPSG covers
https://spatialreference.org/ref/epsg/2249/

In [14]:
structure_ddf = structure_ddf.to_crs(epsg=2249) 

In [23]:
worcester_boundary.compute()

TOWN  TOWN_ID  TOWNS_ID  POP1980  POP1990  POP2000  POPCH80_90  \
0  WORCESTER      348       261   159843   169759   172648        9916   

   POPCH90_00 TYPE  ISLAND  ... FIPS_PLACE  FIPS_MCD  FIPS_COUNT      ACRES  \
0        2889    C       0  ...      82000  25027030          27  24602.486   

  SQUARE_MIL                              GlobalID  OBJECTID    Shape__Are  \
0    384.411  642fd55c-8273-4e35-993c-d5d0e1eaf706         1  1.071680e+09   

      Shape__Len                                           geometry  
0  139826.751833  POLYGON ((583605.513 2938565.252, 586219.333 2...  

[1 rows x 24 columns]

In [15]:
# Ensure geometry columns are set up correctly for the join

structure_ddf = structure_ddf.set_geometry('geometry')
worcester_boundary = worcester_boundary.set_geometry('geometry')
print(structure_ddf.geometry.name)
print(worcester_boundary.geometry.name)

geometry
geometry


##### If you have a high computing power, you can delete the '#'s for the .compute() code and run to see if the number of strucutres changes.

the 'entire' version compares the entire dataset to the shapefile. This should guarantee that every Worcester building is included.

The 'PROP' version, compares the dataset specifed for the city of Worcester to the shapefile. This is a test to see if the dataset correctly tagged all buildings within the city of Worcester. 

Similar verifications can be done on the state level or other cities, again these are computation heavy though!

In [16]:
worcester_entire_ddf = dg.sjoin(structure_ddf, worcester_boundary, how='inner', predicate='within') # within makes sure the whole structure_ddf 
                                                                                                       # row is inside the worcester_boundary

# Understand this approach may mean some of the structure_ddf rows near the boundary won't be included if the shapefile is off by just a little.

In [ ]:
# worcester_entire_ddf.count().compute()

In [17]:
worcester_PROP_ddf = dg.sjoin(worcester_PROP_CITY_ddf, worcester_boundary, how='inner', predicate='within')

In [ ]:
# worcester_PROP_ddf.count().compute()

# THIS CAN BE NEXT-STEPS // FURTHER FUTURE ANALYSIS. It does not need to be completed 100%
Review Census data within Worcester. Grab the shapefiles for two of the census blocks. spatial filter & intersection with the dataset again.

Then review and draw relationships from the data

##### Also, create a function to download country wide census data (depending on the size. Maybe just do cities/Massachusetts and just provide a link)

# Potential Analysis Examples:

### Review Census data within a town (Hope is less data than Worcester that will allow easier computations)

In [18]:
norwood_ddf = massachusetts_ddf[massachusetts_ddf['PROP_ZIP'] == '02062']

Calculate the square footage for each building and make this stored in a new column 'structure_size'.

In [19]:
norwood_ddf['structure_size'] = norwood_ddf.geometry.area

##### WARNING: Below code also requires extensive computing power

In [ ]:
# Find the average structure size

# norwood_ddf['structure_size'].mean().compute()

In [22]:
bldg_norwood_sorted = norwood_ddf.sort_values(by='structure_size', ascending=False).compute()

# binLength = len(bldg_norwood_sorted) // 3000
plt.hist(bldg_norwood_sorted['structure_size'], bins = 600, edgecolor='black')
plt.xlim([0, 400])
plt.xlabel('Area (in meters)')
plt.ylabel('Frequency')
plt.title('area_in_meters for Norwood Buildings')
plt.show()

KeyboardInterrupt: 

In [ ]:
norwood_ddf.shape[0].compute() # Number of buildings that are + 450 sqft

In [ ]:
# Could get the shapefile for Norwood and perform these